internal/walk/walklinks.go

```diff
@@ -23,19 +23,33 @@ func pathRewriter(src string, dest string, fn fs.WalkDirFunc) fs.WalkDirFunc {
 
 // Determines if path matches pattern, striving for parity with rsync,
 // ("Include/Exclude Pattern Rules", https://linux.die.net/man/1/rsync).
-func matchPattern(path string, pattern string) (bool, error) {
-	if strings.ContainsAny(pattern, "*?") {
+func matchPattern(path string, pattern string) (result bool, err error) {
+	if pattern == ".*" {
```

Comment:
This should not be needed.  If this is needed it's an immediate red flag that something in the remainder of the algorithm is wrong.

In this case I believe what's wrong is the fact that *other* patterns are being applied without ^ used to anchor them.  It's wrong to think there is a special case for this pattern, instead the handling of *other* patterns is already wrong and you're working around it just for this pattern.

Can you please compare the behavior against rsync yourself?  When I do that, I can immediately see that some of the expected behavior written in the test is wrong.  For example, this case in TestWalkLinksMatchPattern:

```
{"test.txt", `.txt`, true}
```

It says that .txt pattern should match test.txt, but if you try with rsync it in fact does *not* match:

```
$ tree src/
src/
├── home
│   └── file-in-home
├── some
│   ├── dir
│   │   └── link-to-somefile -> ../somefile
│   └── somefile
└── test.txt

3 directories, 4 files

# .txt does not exclude test.txt in rsync
$ /usr/bin/rsync -aviz --dry-run --exclude .txt src/ dest/
sending incremental file list
created directory dest
cd+++++++++ ./
>f+++++++++ test.txt
cd+++++++++ home/
>f+++++++++ home/file-in-home
cd+++++++++ some/
>f+++++++++ some/somefile
cd+++++++++ some/dir/
cL+++++++++ some/dir/link-to-somefile -> ../somefile

sent 314 bytes  received 70 bytes  768.00 bytes/sec
total size is 11  speedup is 0.03 (DRY RUN)

# *.txt on the other hand will exclude it
$ /usr/bin/rsync -aviz --dry-run --exclude '*.txt' src/ dest/
sending incremental file list
created directory dest
cd+++++++++ ./
cd+++++++++ home/
>f+++++++++ home/file-in-home
cd+++++++++ some/
>f+++++++++ some/somefile
cd+++++++++ some/dir/
cL+++++++++ some/dir/link-to-somefile -> ../somefile

sent 282 bytes  received 67 bytes  698.00 bytes/sec
total size is 11  speedup is 0.03 (DRY RUN)
```

I know the JIRA you're working on specifically says that ".*" doesn't work correctly, but we should not have to special case just that one pattern to make it work.  That would leave many other patterns still broken.  Could you please look at it from a higher level and get it working more consistently with rsync in general?

Reply:
Instead of having to manually come up with these reproducers, here is a script which can be used:

```
$ cat ~/bin/rsync-exclude-test 
#!/bin/sh
set -o pipefail
set -e

WORKDIR=$(mktemp --tmpdir -d rsync_testXXX)
trap 'rm -rf $WORKDIR' EXIT

pattern="$1"
path="$2"

mkdir -p "$WORKDIR/src/$(dirname $path)"
mkdir -p "$WORKDIR/dest"

touch "$WORKDIR/src/$path"

/usr/bin/rsync -aiz --exclude "$pattern" $WORKDIR/src/ $WORKDIR/dest >/dev/null

if test -f "$WORKDIR/dest/$path"; then
  echo "rsync exclude '$pattern' does NOT exclude '$path'"
else
  echo "rsync exclude '$pattern' excludes '$path'"
fi
```

Examples:

```
$ rsync-exclude-test '.*' test.txt
rsync exclude '.*' does NOT exclude 'test.txt'
$ rsync-exclude-test '*.txt' test.txt
rsync exclude '*.txt' excludes 'test.txt'
$ rsync-exclude-test '.*' foo/bar/baz/quux.txt
rsync exclude '.*' does NOT exclude 'foo/bar/baz/quux.txt'
$ rsync-exclude-test '.*' foo/bar/.baz/quux.txt
rsync exclude '.*' excludes 'foo/bar/.baz/quux.txt'
```

Our exclude handling doesn't have to support every possible rsync behavior, but generally there should not be differences between exodus-rsync and rsync for simple stuff like the above.

Reply:
Oh, wow. Thanks. The script was very helpful in redefining test cases to ensure parity.